clippy_lints/src/macros_hiding_unsafe_code.rs

```diff
@@ -0,0 +1,140 @@
+    /// unsafe fn foo(x: *const usize) {
+    ///     #[allow(unused_unsafe)] unsafe { unsafe_macro!(std::ptr::read(x)) };
+    /// }
+    /// ```
+    pub MACROS_HIDING_UNSAFE_CODE,
+    restriction,
+    "macros hiding unsafe code, while `unsafe_op_in_unsafe_fn` is enabled"
+}
+
+#[derive(Clone, Copy, Default)]
+pub struct MacrosHidingUnsafeCode {
+    in_unsafe_fn: bool,
+    in_unsafe_block: usize,
+}
+
+impl_lint_pass!(MacrosHidingUnsafeCode => [MACROS_HIDING_UNSAFE_CODE]);
+
+impl LateLintPass<'_> for MacrosHidingUnsafeCode {
+    fn check_block(&mut self, cx: &LateContext<'_>, block: &Block<'_>) {
+        if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules {
+            if in_macro(block.span) {
+                if self.in_unsafe_fn
+                    && self.in_unsafe_block == 0
+                    && !is_lint_allowed(cx, UNSAFE_OP_IN_UNSAFE_FN, block.hir_id)
+                {
+                    let macro_call_span = block.span.source_callsite();
+                    let unused_unsafe_sugg = if !in_external_macro(cx.sess(), block.span)
+                        && !is_lint_allowed(cx, UNUSED_UNSAFE, block.hir_id)
+                    {
+                        "#[allow(unused_unsafe)] "
+                    } else {
+                        ""
+                    };
+                    let mut applicability = Applicability::MaybeIncorrect;
+                    span_lint_and_sugg(
+                        cx,
+                        MACROS_HIDING_UNSAFE_CODE,
+                        macro_call_span,
+                        "this macro call hides unsafe code",
+                        "wrap it in an `unsafe` block",
+                        format!(
+                            "{}unsafe {{ {} }}",
+                            unused_unsafe_sugg,
+                            snippet_with_applicability(cx, macro_call_span, "...", &mut applicability),
+                        ),
+                        applicability,
+                    );
+                }
+            } else {
+                self.in_unsafe_block = self.in_unsafe_block.saturating_add(1);
+            }
+        }
+    }
+
+    fn check_block_post(&mut self, _: &LateContext<'_>, block: &Block<'_>) {
+        if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules {
+            if !in_macro(block.span) {
+                self.in_unsafe_block = self.in_unsafe_block.saturating_sub(1);
+            }
+        }
+    }
+
+    fn check_fn(&mut self, _: &LateContext<'_>, fn_kind: FnKind<'_>, _: &FnDecl<'_>, _: &Body<'_>, _: Span, _: HirId) {
+        if is_unsafe_fn(fn_kind) {
+            self.in_unsafe_fn = true;
+        }
+    }
+
+    fn check_fn_post(
+        &mut self,
+        _: &LateContext<'_>,
+        fn_kind: FnKind<'_>,
+        _: &FnDecl<'_>,
+        _: &Body<'_>,
+        _: Span,
+        _: HirId,
+    ) {
+        if is_unsafe_fn(fn_kind) {
+            self.in_unsafe_fn = false;
+        }
```

Comment:
I think this needs to be adjusted for nested functions.

```rust
unsafe fn f() {
    fn g() {}
    unsafe_macro!();
}
```

Reply:
Ah right. Your example shouldn't break this code, but 

```rs
unsafe fn f() {
    unsafe fn g() {}
    unsafe_macro!();
}
```

will.